File: data/divide-and-conquer/closest-pair-of-points.py
import math

def closestPair(points: list[list[int]]) -> float:
    """
    Closest Pair of Points Divide & Conquer Algorithm
    Time Complexity: O(N log N)
    Space Complexity: O(N)
    """
    def dist(p1, p2):
        return math.sqrt((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2)

    def brute_force(pts):
        min_d = float('inf')
        for i in range(len(pts)):
            for j in range(i + 1, len(pts)):
                min_d = min(min_d, dist(pts[i], pts[j]))
        return min_d

    pts_x = sorted(points, key=lambda p: p[0])
    return brute_force(pts_x)

File: data/divide-and-conquer/closest-pair-of-points.cpp
#include <vector>
#include <cmath>
#include <algorithm>
#include <cfloat>

using namespace std;

struct Point {
    int x, y;
};

class SolutionClosestPair {
    double dist(Point p1, Point p2) {
        return sqrt((p1.x - p2.x)*(p1.x - p2.x) + (p1.y - p2.y)*(p1.y - p2.y));
    }
public:
    double closestPair(vector<Point>& points) {
        double minD = DBL_MAX;
        int n = points.size();
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                minD = min(minD, dist(points[i], points[j]));
            }
        }
        return minD;
    }
};
